lib/Dialect/HW/HWAttributes.cpp

```diff
@@ -48,6 +51,108 @@ void HWDialect::printAttribute(Attribute attr, DialectAsmPrinter &p) const {
   llvm_unreachable("Unexpected attribute");
 }
 
+//===----------------------------------------------------------------------===//
+// OutputFileAttr
+//===----------------------------------------------------------------------===//
+
+static std::pair<std::string, std::string>
+normalizeFilename(StringRef directory, StringRef filename) {
+  SmallString<128> fullpath;
+
+  // If the filename is an absolute path, we don't need the working directory.
+  if (llvm::sys::path::is_absolute(filename))
+    fullpath = filename;
+  else
+    llvm::sys::path::append(fullpath, directory, filename);
+
+  // Remove any redundant path parts.
+  llvm::sys::path::remove_dots(fullpath, /*remove_dot_dot=*/true);
+
+  // If the fullpath ends in a `/` we consider this to be a directory target
+  // with an empty filename.
+  if (fullpath.endswith("/")) {
+    // We strip the trailing slash off the directory name, unless its the root
+    // directory which is only a `/`.
+    if (fullpath.size() > 1)
+      fullpath.pop_back();
+    return std::make_pair(std::string(fullpath), "");
+  }
+
+  // Split the file name into directory + file.  If the full path is a
+  // directory, the filename will just be a '/'.
+  filename = llvm::sys::path::filename(fullpath);
+  directory = llvm::sys::path::parent_path(fullpath);
+  return std::pair(directory.str(), filename.str());
+}
```

Comment:
Nice use of `llvm::sys::path`! I originally was worried that we were rolling our own stuff here, but this looks fantastic. :100: